insights/Extract_MC_Insights_DBX.py

```diff
@@ -0,0 +1,73 @@
+# Databricks notebook source
+# MAGIC %md # Environment Setup
+
+# COMMAND ----------
+
+#Only run this if you have not installed pycarlo on the DBX cluster already
+#pip install pycarlo
+
+# COMMAND ----------
+
+# DBTITLE 1,Runtime Variables
+# Preferred: Monte Carlo Credentials stored in DBX Secret Key Repo called "monte-carlo-creds":
+#mcd_id = dbutils.secrets.get(scope = "monte-carlo-creds", key = "mc-key-id")
+#mcd_token = dbutils.secrets.get(scope = "monte-carlo-creds", key = "mc-secret-id")
+
+#Otherwise hardcode for testing purposes only!
+mcd_id = 'your-id-here'
+mcd_token = 'your-secret-here'
+
+# Full List of Reports Needed, Edit as Needed for Databricks
+insight_names = ["monitors","cleanup_suggestions","events","incident_history"]
+insight_report_names = ["monitors.csv","cleanup_suggestions.csv","events.csv","incident_history.csv"]
```

Comment:
If you prefer you can retrieve a list of all available reports from the [getInsights API](https://apidocs.getmontecarlo.com/#query-getInsights).

Here is a quick example of doing so and uploading to S3 (can also use that a source for generating tables if you'd like):
```
from pathlib import Path
from typing import Optional

import boto3
import requests

from pycarlo.core import Client, Query

MC_CLIENT = Client()
S3_CLIENT = boto3.client('s3')


def upload_insights_to_s3(destination_bucket: str, desired_file_extension: Optional[str] = '.csv') -> None:
    """Example function for listing all insights in an account, and uploading any available to S3 as a CSV."""
    list_insights_query = Query()
    list_insights_query.get_insights()
    for insight in MC_CLIENT(list_insights_query).get_insights:
        report_name = str(Path(insight.name).with_suffix(desired_file_extension))

        if insight.available:
            report_url_query = Query()
            report_url_query.get_report_url(insight_name=insight.name, report_name=report_name)
            report_url = MC_CLIENT(report_url_query).get_report_url.url

            print(f'Uploading {report_name} to {destination_bucket}.')
            S3_CLIENT.upload_fileobj(
                Fileobj=requests.get(url=report_url, stream=True).raw,
                Bucket=destination_bucket,
                Key=report_name
            )


if __name__ == "__main__":
    upload_insights_to_s3(destination_bucket='<BUCKET-NAME>')
```

Reply:
Thanks @pxc-dev  - I am not going to use this approach as it gives errors at runtime. Given that DBX based connections only allow specific reports, I am going to leave it at the hardcoded list which users can modify as needed.
`Uploading field_cleanup_suggestions.csv to {buckey_name}.
{path}/monte_carlo/field_cleanup_suggestions.csv
MissingSchema: Invalid URL 'None': No scheme supplied. Perhaps you meant https://None?`

Reply:
Are those insights not marked as `available=false`, which should be skipped by that "if block"?

Reply:
When I try this in our DBX account, I get schema errors in python because the resulting HTTP URL result is empty (or set to "None".)  It seems to happen with "field_cleanup_suggestions.csv" which is available in the MC UI but seems to be erroring here.

Reply:
This fixes the error and bypasses it:
`if str(report_url) == 'None':
                print('An error getting report: '+report_name)
            else:
                S3_CLIENT.upload_fileobj(
                    Fileobj=requests.get(url=report_url, stream=True).raw,
                    Bucket=destination_bucket,
                    Key=filepath+report_name
                )`